odoo/modules/loading.py

```diff
@@ -419,6 +419,9 @@ def load_modules(registry, force_demo=False, status=None, update_module=False):
         if not graph:
             _logger.critical('module base cannot be loaded! (hint: verify addons-path)')
             raise ImportError('Module `base` cannot be loaded! (hint: verify addons-path)')
+        if update_module and tools.config['update']:
+            for pyfile in tools.config['pre_upgrade_scripts'].split(','):
+                odoo.modules.migration.exec_script(cr, graph['base'].installed_version, pyfile, 'base', 'pre')
```

Comment:
Be careful about this logic after my `graph.py` refactoring.
Not sure if it would be a corner case:
After my refactor the `tools.config['update']` is static and won't be cleared after upgrade
if the upgrade script will generate `to remove` modules. The registry will be loaded again with `module_update=True`.
https://github.com/odoo/odoo/blob/c4c2a30effd3e918001a8e41b7c2df61aebd1216/odoo/modules/loading.py#L535
which means the extra upgrade script will be executed twice!

Reply:
I guess we will figure this out once we reach that point in the forward port. You will be in the loop anyway :) Thanks for heads up.